### DISCUSSION TITLE
PID implementation, questions about derivative term

### DISCUSSION THREAD

**Discussion by GusBon** [bitcraze/discussions#1420] [Q&A] [answered]
Hi,

I am currently reading the documentation and the firmware code to understand better the different part of the code and control principle behind.
So reading the PID part of the code, I understood that the PID structure is implemented in pid.c file, and I have some questions about how the derivative part of the PID is implemented. The following link refers to the part of the code that disturbs me:
https://github.com/bitcraze/crazyflie-firmware/blob/7cbe7cc56ff52d9b0b7f9442858d801eca35a6bd/src/utils/src/pid.c#L57-L123

And more specifically those lines:
```C
float deriv = -(measured - pid->prevMeasured) / pid->dt;
  
  #if CONFIG_CONTROLLER_PID_FILTER_ALL
    pid->deriv = deriv;
  #else
    if (pid->enableDFilter){
      pid->deriv = lpf2pApply(&pid->dFilter, deriv);
    } else {
      pid->deriv = deriv;
    }
  #endif
  if (isnan(pid->deriv)) {
    pid->deriv = 0;
  }
  pid->outD = pid->kd * pid->deriv;
  output += pid->outD;
```
I don't understand why the reference is set to 0 when calculating the "deriv" term, like the desired part of the error is not taken in account. Is it a common thing ? I would probably implement something like:
```C
float deriv = (pid->desired pid->prevDesired) / pid->dt -(measured - pid->prevMeasured) / pid->dt;
```

By looking at the gains in the platform_defaults_cf2.h, a lot of PID don't actually use the D term so this might not change a lot, but if there is an explanation to this specific implementation I would be glad to learn it !

Thanks in advance !

**Comment by knmcguire** [comment]
Hi!

Are you talking about this part?

```
  if (isnan(pid->deriv)) {
    pid->deriv = 0;
  }
```

I think it was just a quick solution to deal with any NaN numbers since a operation with that will deal with NaN later on the road will reduce in run errors. Perhaps that is some kind of case that was not handled in the lp2p?

In terms of the way that you are suggestion, the derivative term is referring to the derivative to the error is supposed to be this indeed:

```
error =  desired - measured
derivative error = current_error - previous_error / dt
```

Why it was implemented like how it is now... who knows. It's been a very long time ago and I guess a lot of things were done to get it fly at this point and no-one really encountered that issue since the gain wasn't non-zero. And in a sense, the current implementation will have the dampening effect as well, but perhaps not a theoretical correct on. 

We should probably make an issue out of this :) Thanks for letting us know

**Reply by knmcguire** [reply, marked as answer]
Ah, this was changed a while ago with this PR (which I merged btw.... forgot about that haha), to prevent the derivative kick with setpoint changes. https://github.com/bitcraze/crazyflie-firmware/issues/1336.

So technically it was a implications based that the setpoint doesn't change that much over time.

**Reply by GusBon** [reply]
Hi !

Thanks for the quick response !

After looking at automatic lessons it is normal and a good thing to implement numerical PID this way, as this avoid the derivative kick and still play the role of a damper. I should have done this before asking.

Sorry for the disturbance and thanks again !
